lms/djangoapps/grades/course_grade_factory.py

```diff
@@ -20,48 +20,33 @@ class CourseGradeFactory(object):
     """
     GradeResult = namedtuple('GradeResult', ['student', 'course_grade', 'error'])
 
-    def create(self, user, course=None, collected_block_structure=None, course_structure=None, course_key=None):
+    def read(
+            self,
+            user,
+            course=None,
+            collected_block_structure=None,
+            course_structure=None,
+            course_key=None,
+            create_if_needed=True,
+    ):
         """
         Returns the CourseGrade for the given user in the course.
-        Reads the value from storage and validates that the grading
-        policy hasn't changed since the grade was last computed.
+        Reads the value from storage.
         If not in storage, returns a ZeroGrade if ASSUME_ZERO_GRADE_IF_ABSENT.
-        Else, if changed or not in storage, computes and returns a new value.
-
-        At least one of course, collected_block_structure, course_structure,
-        or course_key should be provided.
-        """
-        course_data = CourseData(user, course, collected_block_structure, course_structure, course_key)
-        try:
-            course_grade, read_policy_hash = self._read(user, course_data)
-            if read_policy_hash == course_data.grading_policy_hash:
-                return course_grade
-            read_only = False  # update the persisted grade since the policy changed; TODO(TNL-6786) remove soon
-        except PersistentCourseGrade.DoesNotExist:
-            if assume_zero_if_absent(course_data.course_key):
-                return self._create_zero(user, course_data)
-            read_only = True  # keep the grade un-persisted; TODO(TNL-6786) remove once all grades are backfilled
-
-        return self._update(user, course_data, read_only)
-
-    def read(self, user, course=None, collected_block_structure=None, course_structure=None, course_key=None):
-        """
-        Returns the CourseGrade for the given user in the course as
-        persisted in storage.  Does NOT verify whether the grading
-        policy is still valid since the grade was last computed.
-        If not in storage, returns a ZeroGrade if ASSUME_ZERO_GRADE_IF_ABSENT
-        else returns None.
+        Else if create_if_needed, computes and returns a new value.
```

Comment:
Would you mind writing out (not in the docstring, just in a comment here on the review) why ASSUME_ZERO_GRADE_IF_ABSENT takes precedence over create_if_needed?

Reply:
Good question.  For our environment with both persistent-grades and assume-zero-if-absent enabled, we would never hit this else-if and else clauses.  However, currently, in the open-edx environments where this is not the case, they will need to recompute the grades on-demand.  However, for certain views (like the student dashboard), we want to avoid this heavy computation in open-edx (and just not show the grades at all).  Currently, the student dashboard view is the only one that passes `create_if_needed=False`.

Going forward, if there are additional features that rely on persistent-grades for scalability, we'll need to watch out for the impact in the open-edx environment.

Reply:
Thanks!